In [6]:
import pandas as pd
import numpy as np
import os

# Raw - Mobility & Air quality

Importación de datos de calidad de aire y mobilidad

inicialmente se leen los datos crudos desde los archivos descargables de las  respectivas fuentes, se procede a compilar todos los datos de calidad de aire en 3 dataframe diferentes (Train, Test y Validation) que se convierten posteriormente a un archivo HDF5 el cual permite recolectar la información de manera más eficiente facilitando la lectura para proximas ocasiones, de igual maneraa se procede con los datos de mobilidad 

In [7]:
#"""
directory_path = 'C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train'
dfs = []

for filename in os.listdir(directory_path):
    if filename.endswith('.csv'):
        file_path = os.path.join(directory_path, filename)
        df = pd.read_csv(file_path)
        dfs.append(df)

Train_air = pd.concat(dfs, ignore_index=True)

# Save the combined DataFrame to an HDF5 file
Train_air.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df', mode='w') 
#"""

In [8]:
# Load the DataFrame from the HDF5 file
Train_air = pd.read_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df')

Los datos de train van desde 2021-01-01 hasta 2023-05-01

In [9]:
Train_air.tail()

Fecha_Hora  codigoSerial     pm25  calidad_pm25    pm10  \
194035  2024-02-29 19:00:00            86  27.7981           1.0 -9999.0   
194036  2024-02-29 20:00:00            86  32.4447           1.0 -9999.0   
194037  2024-02-29 21:00:00            86  34.7520           1.0 -9999.0   
194038  2024-02-29 22:00:00            86  47.7134           1.0 -9999.0   
194039  2024-02-29 23:00:00            86  55.8835           1.0 -9999.0   

        calidad_pm10     pm1  calidad_pm1      no  calidad_no  ...   p_ssr  \
194035           1.0 -9999.0          1.0 -9999.0         1.0  ... -9999.0   
194036           1.0 -9999.0          1.0 -9999.0         1.0  ... -9999.0   
194037           1.0 -9999.0          1.0 -9999.0         1.0  ... -9999.0   
194038           1.0 -9999.0          1.0 -9999.0         1.0  ... -9999.0   
194039           1.0 -9999.0          1.0 -9999.0         1.0  ... -9999.0   

        calidad_p_ssr  pliquida_ssr  calidad_pliquida_ssr  rglobal_ssr  \
194035            1.0       -9999.0                   1.0      -9999.0   
194036            1.0       -9999.0                   1.0      -9999.0   
194037            1.0       -9999.0                   1.0      -9999.0   
194038            1.0       -9999.0                   1.0      -9999.0   
194039            1.0       -9999.0                   1.0      -9999.0   

        calidad_rglobal_ssr  taire10_ssr  calidad_taire10_ssr  vviento_ssr  \
194035                  1.0      -9999.0                  1.0      -9999.0   
194036                  1.0      -9999.0                  1.0      -9999.0   
194037                  1.0      -9999.0                  1.0      -9999.0   
194038                  1.0      -9999.0                  1.0      -9999.0   
194039                  1.0      -9999.0                  1.0      -9999.0   

        calidad_vviento_ssr  
194035                  1.0  
194036                  1.0  
194037                  1.0  
194038                  1.0  
194039                  1.0  

[5 rows x 36 columns]

In [10]:
Train_air.columns

Index(['Fecha_Hora', 'codigoSerial', 'pm25', 'calidad_pm25', 'pm10',
       'calidad_pm10', 'pm1', 'calidad_pm1', 'no', 'calidad_no', 'no2',
       'calidad_no2', 'nox', 'calidad_nox', 'ozono', 'calidad_ozono', 'co',
       'calidad_co', 'so2', 'calidad_so2', 'pst', 'calidad_pst', 'dviento_ssr',
       'calidad_dviento_ssr', 'haire10_ssr', 'calidad_haire10_ssr', 'p_ssr',
       'calidad_p_ssr', 'pliquida_ssr', 'calidad_pliquida_ssr', 'rglobal_ssr',
       'calidad_rglobal_ssr', 'taire10_ssr', 'calidad_taire10_ssr',
       'vviento_ssr', 'calidad_vviento_ssr'],
      dtype='object')

In [11]:
Air = Train_air.copy()
estaciones_eliminar = [79,80,12] # se eliminan para tener la linea de tiempo de un solo sensor y no repetir fechas
Train_air = Train_air[~Train_air['codigoSerial'].isin(estaciones_eliminar)]

In [12]:
Air = Train_air[['Fecha_Hora', 'codigoSerial', 'pm25', 'calidad_pm25']]

Del dataset anterior únicamente se tomaran en cuenta las columnas de fecha_hora, codigoSerial, pm25 y calidad_pm25, ya que, las demás columnas al tener valores de -9999 y calidad 1 significan que los sensores están fuera de servicio y por la descripción del dataset se evidencia que solo las columnas de pm25 tienen un comportamiento diferente al mencionado

In [13]:
#Ponemos bajo un mismo formato a la columnda de Fecha_Hora
Air.loc[:,'Fecha_Hora'] = pd.to_datetime(Air['Fecha_Hora'], format='%Y-%m-%d %H:%M:%S')

In [14]:
# Establecer la columna 'fecha' como índice
Air.set_index('Fecha_Hora', inplace=True)

C:\Users\MARIA PAULA\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\pandas\core\indexes\base.py:7588: FutureWarning: Dtype inference on a pandas object (Series, Index, ExtensionArray) is deprecated. The Index constructor will keep the original dtype in the future. Call `infer_objects` on the result to get the old behavior.
  return Index(sequences[0], name=names)


In [15]:
Air.head(10)

codigoSerial     pm25  calidad_pm25
Fecha_Hora                                              
2021-01-01 00:00:00            83  23.7741           1.0
2021-01-01 01:00:00            83  28.5095           1.0
2021-01-01 02:00:00            83  27.2400           1.0
2021-01-01 03:00:00            83  37.8129           1.0
2021-01-01 04:00:00            83  34.1848           1.0
2021-01-01 05:00:00            83  34.1002           1.0
2021-01-01 06:00:00            83  45.9056           1.0
2021-01-01 07:00:00            83  28.2544           1.0
2021-01-01 08:00:00            83  23.7000           1.0
2021-01-01 09:00:00            83  20.2378           1.0

La columna codigoSerial representa el código de la estación con esta podemos saber donde se encuentra ubicada, para poder comparar los datos con el dataset de mobilidad

Serial 83   
Comuna 16
Belén - I.E Pedro Justo Berrio

Serial 84
Comuna 14
El Poblado - I.E INEM Sede Santa Catalina

Serial 85
Comuna 10
CEN-TRAF - San Cristobal - Parque Biblioteca Fernando Botero

Serial 86
Comuna 4
Aranjuez - I.E Ciro Mendia

https://www.medellin.gov.co/es/conoce-algunos-datos-generales-de-la-ciudad/


Creamos una función para calcular el porcentaje de datos nulos en el dataset

In [16]:
def missing_values(df):
    Column_Names = df.columns

    Total_rows = df.shape[0]

    missing_values = df.isnull().sum()

    Percent_missing = (missing_values / Total_rows)*100
    
    result = pd.DataFrame({
        'Column_Name': Column_Names,
        'Missing_Values': missing_values.values,
        'Total_Rows': Total_rows,
        'Percent_Missing': Percent_missing.values
    })

    return result

Cambiamos la distribución de la tabla con ayuda de una tabla pivote, así por cada una de las estaciones tenemos una columna para calidad y otra para la magnitud de pm2.5

In [17]:
Air_organized = Air.pivot_table(index=Air.index, columns='codigoSerial', values=['pm25', 'calidad_pm25'], aggfunc = 'mean')

In [18]:
Air_organized.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 27720 entries, 2021-01-01 00:00:00 to 2024-02-29 23:00:00
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   (calidad_pm25, 83)  27720 non-null  float64
 1   (calidad_pm25, 84)  27720 non-null  float64
 2   (calidad_pm25, 85)  27720 non-null  float64
 3   (calidad_pm25, 86)  27720 non-null  float64
 4   (pm25, 83)          27720 non-null  float64
 5   (pm25, 84)          27720 non-null  float64
 6   (pm25, 85)          27720 non-null  float64
 7   (pm25, 86)          27720 non-null  float64
dtypes: float64(8)
memory usage: 1.9 MB


In [19]:
missing_values(Air_organized)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0  (calidad_pm25, 83)               0       27720              0.0
1  (calidad_pm25, 84)               0       27720              0.0
2  (calidad_pm25, 85)               0       27720              0.0
3  (calidad_pm25, 86)               0       27720              0.0
4          (pm25, 83)               0       27720              0.0
5          (pm25, 84)               0       27720              0.0
6          (pm25, 85)               0       27720              0.0
7          (pm25, 86)               0       27720              0.0

Resulta que este dataset tiene la particularidad que los datos nulos o erroneos no son NaN, sino que se conocen por el valor de la calidad del datos, si es de1 el valor es real y ha sido revisado, si es un numero diferente, puede ser dudoso, malo, el sensor estar fuera de servicio, por lo que unicamente tomaremos los datos con valores de calidad de 2.6

In [20]:
# Revisamos la cantidad de datos con diferentes calificaciones de calidad
# Seleccionar todas las columnas con el nombre 'calidad_pm25' en el primer nivel del MultiIndex
calidad_pm25_columns = Air_organized.xs('calidad_pm25', level=0, axis=1)

# Iterar sobre cada columna (por código serial)
for column in calidad_pm25_columns.columns:
    # Contar la frecuencia de cada valor único
    value_counts = calidad_pm25_columns[column].value_counts()
    
    # Mostrar los resultados
    print(f"Columna (código serial): {column}")
    print(value_counts)
    print("\n")


Columna (código serial): 83
83
1.00    26240
2.30      379
4.31      311
3.10      216
4.32      143
3.00      140
2.20      111
3.30      109
2.40       67
3.20        4
Name: count, dtype: int64


Columna (código serial): 84
84
1.00    25867
3.10      429
4.31      302
2.30      289
3.00      272
3.30      153
2.40      144
4.32      140
2.20      118
3.20        6
Name: count, dtype: int64


Columna (código serial): 85
85
1.00    25694
4.31      493
2.30      379
3.10      359
4.32      241
3.30      176
2.20      156
3.00      144
2.40       75
3.20        3
Name: count, dtype: int64


Columna (código serial): 86
86
1.00    26015
2.30      449
4.31      368
3.10      306
3.30      215
4.32      138
3.00      108
2.20       92
2.40       23
3.20        6
Name: count, dtype: int64




In [22]:
Air_organized.describe().T

count        mean          std     min  \
             codigoSerial                                             
calidad_pm25 83            27720.0    1.116052     0.525356     1.0   
             84            27720.0    1.144059     0.570935     1.0   
             85            27720.0    1.168476     0.640962     1.0   
             86            27720.0    1.135961     0.567426     1.0   
pm25         83            27720.0  427.243486  7287.595887 -9999.0   
             84            27720.0  460.474039  7422.232152 -9999.0   
             85            27720.0  721.881973  9472.277093 -9999.0   
             86            27720.0  408.849149  7318.296096 -9999.0   

                                 25%       50%        75%        max  
             codigoSerial                                             
calidad_pm25 83             1.000000   1.00000   1.000000       4.32  
             84             1.000000   1.00000   1.000000       4.32  
             85             1.000000   1.00000   1.000000       4.32  
             86             1.000000   1.00000   1.000000       4.32  
pm25         83            10.774600  17.98150  27.400025  117237.00  
             84             9.590720  14.74290  20.808350  117429.00  
             85             9.831675  14.67145  20.541575  100117.00  
             86            11.846975  17.01035  23.408200  108778.00

Desde el diccionario de datos del SIATA menciona que valores de calidad mayores a 2.6 no son confiables por lo que se eliminan de mi dataset en aras de mejorar la calidad de mis datos y por ende la predicción

In [ ]:
prueba = Air_organized.copy()

# Filtrar todas las columnas de 'calidad_pm25'
calidad_columns = prueba.xs('calidad_pm25', level=0, axis=1)

# Iterar sobre cada código serial en las columnas 'calidad_pm25'
for codigoSerial in calidad_columns:
    # Reemplazar los valores en 'pm25' si 'calidad_pm25' es menor a 1
    prueba.loc[prueba[('calidad_pm25', codigoSerial)] > 1, ('pm25', codigoSerial)] = np.nan
    #prueba.loc[prueba[('calidad_pm25', codigoSerial)] > 1, ('calidad_pm25', codigoSerial)] = np.nan


In [24]:
missing_values(prueba)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0  (calidad_pm25, 83)               0       27720         0.000000
1  (calidad_pm25, 84)               0       27720         0.000000
2  (calidad_pm25, 85)               0       27720         0.000000
3  (calidad_pm25, 86)               0       27720         0.000000
4          (pm25, 83)            1480       27720         5.339105
5          (pm25, 84)            1853       27720         6.684704
6          (pm25, 85)            2026       27720         7.308802
7          (pm25, 86)            1705       27720         6.150794

In [25]:
# Crear columnas auxiliares de año, mes, día a partir del índice
prueba['Año'] = prueba.index.year
prueba['Mes'] = prueba.index.month
prueba['Día'] = prueba.index.day

# Calcular la media horaria para cada día específico
# Esto incluye agrupar por año, mes, día  y luego calcular la media
medias_horarias = prueba.groupby(['Año', 'Mes', 'Día']).transform('mean')

# Rellenar valores faltantes usando las medias horarias calculadas
prueba.fillna(medias_horarias, inplace=True)

# Eliminar las columnas auxiliares si no son necesarias
prueba.drop(['Año', 'Mes', 'Día'], axis=1, inplace=True)

In [26]:
missing_values(prueba)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0  (calidad_pm25, 83)               0       27720         0.000000
1  (calidad_pm25, 84)               0       27720         0.000000
2  (calidad_pm25, 85)               0       27720         0.000000
3  (calidad_pm25, 86)               0       27720         0.000000
4          (pm25, 83)             192       27720         0.692641
5          (pm25, 84)             240       27720         0.865801
6          (pm25, 85)             288       27720         1.038961
7          (pm25, 86)             168       27720         0.606061

Al ya no necesitar los datos de calidad_pm25 pues hemos ya sacado/borrado las magnitudes asociadas a aquellas calidades diferentes de 1, sólo sacamos la descripción de las magnitudes por cada estación

In [27]:
prueba['pm25'].describe().T

count       mean        std  min        25%       50%  \
codigoSerial                                                            
83            27528.0  19.924938  11.406194  0.0  11.092275  18.14665   
84            27480.0  15.934843   8.259910  0.0  10.002525  14.83860   
85            27432.0  15.731141   7.595484  0.0  10.221475  14.73365   
86            27552.0  18.276499   8.547930  0.0  12.137750  17.39395   

                    75%       max  
codigoSerial                       
83            27.035250   78.2079  
84            20.486800   75.6756  
85            19.815375   65.9214  
86            22.968244  110.5340

Para los datos nulos, se tienen 2 alternativas

1. **Forward or backward fill:** Use the value from the previous or next observation to fill null values. This method is especially useful for time series data.

df.fillna(method='ffill', inplace=True, axis=0)

2. **Interpolation:** Estimate missing values based on other values in the dataset. Pandas provides various interpolation methods such as linear, polynomial, or spline interpolation.

df.interpolate(method='linear', inplace=True, axis=0)

Inicialmente se llevara a cabo **Forward or backward fill** pues es más acorde al tipo de datos

In [29]:
# Forward fill null values
intento = prueba['pm25'].copy()
intento.ffill(axis=0, inplace=True)


In [31]:
intento.describe().T

count       mean        std  min        25%       50%  \
codigoSerial                                                            
83            27720.0  19.961330  11.384780  0.0  11.115125  18.19500   
84            27720.0  15.896683   8.247713  0.0   9.917191  14.81805   
85            27720.0  15.727549   7.558831  0.0  10.334475  14.73365   
86            27720.0  18.276964   8.522088  0.0  12.165225  17.46520   

                    75%       max  
codigoSerial                       
83            27.107725   78.2079  
84            20.450300   75.6756  
85            19.755175   65.9214  
86            22.856125  110.5340

| Rango de PM2.5 (µg/m³) | Nivel de Salud                     |
|--------------------------|------------------------------------|
| 0 - 50                   | Saludable                          |
| 51 - 100                 | Moderado                           |
| 101 - 150                | Poco saludable para grupos sensibles |
| 151 - 200                | Poco saludable para la población en general |
| Más de 200               | Muy poco saludable o peligroso     |


Para poder unir este dataset con el de mobilidad debo crear una nueva columna que tenga el numero de comuna al cual pertenece cada fila.

12 - comuna 10
79 - comuna 16 
80 - comuna 8
83 - comuna 16
84 - comuna 14
85 - comuna 10
86 - comuna 4

In [32]:
intento.columns.name = None
# Resetear el índice para tener 'Fecha_Hora' como columna
intento.reset_index(inplace=True)
intento.rename(columns={'index': 'Fecha_Hora'}, inplace=True)

# Mapeo de códigos seriales a comunas
mapa_comunas = {12: 10, 79: 16, 80: 8, 83: 16, 84: 14, 85: 10, 86: 4}

# Transformar el dataframe usando melt para desglosar por 'codigoSerial'
df_transformado = intento.melt(id_vars=['Fecha_Hora'], var_name='codigoSerial', value_name='pm_25')

# Crear una columna 'comuna' basada en el código serial
df_transformado['comuna'] = df_transformado['codigoSerial'].map(mapa_comunas)

# Resultado final
print(df_transformado)

                Fecha_Hora codigoSerial    pm_25  comuna
0      2021-01-01 00:00:00           83  23.7741      16
1      2021-01-01 01:00:00           83  28.5095      16
2      2021-01-01 02:00:00           83  27.2400      16
3      2021-01-01 03:00:00           83  37.8129      16
4      2021-01-01 04:00:00           83  34.1848      16
...                    ...          ...      ...     ...
110875 2024-02-29 19:00:00           86  27.7981       4
110876 2024-02-29 20:00:00           86  32.4447       4
110877 2024-02-29 21:00:00           86  34.7520       4
110878 2024-02-29 22:00:00           86  47.7134       4
110879 2024-02-29 23:00:00           86  55.8835       4

[110880 rows x 4 columns]


In [33]:
df_transformado.reset_index(drop=True, inplace=True)  # Esto elimina cualquier índice actual si existe

# Establecer 'Fecha_Hora' como el índice único
df_transformado.set_index('Fecha_Hora', inplace=True)

In [34]:
df_transformado

codigoSerial    pm_25  comuna
Fecha_Hora                                       
2021-01-01 00:00:00           83  23.7741      16
2021-01-01 01:00:00           83  28.5095      16
2021-01-01 02:00:00           83  27.2400      16
2021-01-01 03:00:00           83  37.8129      16
2021-01-01 04:00:00           83  34.1848      16
...                          ...      ...     ...
2024-02-29 19:00:00           86  27.7981       4
2024-02-29 20:00:00           86  32.4447       4
2024-02-29 21:00:00           86  34.7520       4
2024-02-29 22:00:00           86  47.7134       4
2024-02-29 23:00:00           86  55.8835       4

[110880 rows x 3 columns]

In [35]:
df_transformado.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 110880 entries, 2021-01-01 00:00:00 to 2024-02-29 23:00:00
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   codigoSerial  110880 non-null  object 
 1   pm_25         110880 non-null  float64
 2   comuna        110880 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 3.4+ MB


In [36]:
df_final = df_transformado.drop(['codigoSerial'],axis =1)

In [37]:
df_final.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/Train_set_1A.h5', key='df', mode='w', format='table')